### PR TITLE
add cache commands "ls" and "rm"

### PR DESCRIPTION
**Pull request by fritzy**
This adds the ability to list and delete cached package downloads.

`ls [<package>@<version query>]`
Lists all cache keys, or keys related to certain package specs. Filtered results are best effort based on both existing packuments in cache (further resolving package keys) as well as a fallback to filtering tgz packages based on URL.
Any number of packages can be specified.
Packuments cache keys match on any version spec.

`rm [<key>]` (alias of `clean`)
Removes any number of keys

I extended clean to accept any number of cacache keys. Zero arguments still removes everything and requires `--force`.

### CODE REVIEW

File: lib/cache.js [hunk 1]
```diff
@@ -131,6 +162,54 @@ with --force.`)
...
+      if (search.test(key))
+        output.push(key)
+    }
+    return output
+  }
+
+  async list (args) {
+    let pkg, ver
+    if (args.length > 0)
+      [pkg, ver] = args[0].split('@')
```
Comment (ljharb) [inline lib/cache.js]: shouldn't this use https://www.npmjs.com/package/npm-package-arg? the current approach seems like it will break on scoped packages.

**Review by cotaboss — APPROVED** [review]

File: lib/cache.js [hunk 2]
```diff
@@ -29,6 +52,8 @@ class Cache extends BaseCommand {
       'add <tarball url>',
       'add <git url>',
       'add <name>@<version>',
+      'ls',
+      'delete name@<version>',
```
Comment (wraithgar) [inline lib/cache.js]:
```suggestion
      'delete <name>@<version>',
```

File: lib/cache.js [hunk 3]
```diff
@@ -131,6 +162,54 @@ with --force.`)
...
   }
+
+  async searchCacheForPackage (pkg, ver) {
+    const cachePath = path.join(this.npm.cache, '_cacache')
+    const cacheList = await cacache.ls(cachePath)
+    const keyList = Object.keys(cacheList)
+    const output = []
+    if (pkg === undefined)
+      pkg = '(.*)'
+    const search = new RegExp(`^make-fetch-happen:request-cache:(.*)/${pkg}/-/(.*)-(.*).tgz`)
```
Comment (wraithgar) [inline lib/cache.js]: This filename is more fraught with danger than you may realize. Did you know that package names and semver can contain hyphens? It's fun!

File: lib/cache.js (hunk 1 above)
Comment (fritzy) [reply lib/cache.js]: Yup. There's a bunch of changes needed here. This was a first pass POC. My first time getting hands on the codebase. I should have made that clear in my PR. The logic largely needs to move to cacache, the regexp needs some work, and you're absolutely right about `npm-package-arg`.

**Comment by fritzy** [comment]
Additionally I fixed unscoped packages from matching scoped packages in `npm cache ls --package name`

**Comment by fritzy** [comment]
I've finished the test coverage, so this is now ready for review.

File: lib/cache.js [hunk 4]
```diff
@@ -4,7 +4,25 @@ const log = require('npmlog')
...
 const rimraf = promisify(require('rimraf'))
+const semver = require('semver')
 const BaseCommand = require('./base-command.js')
+const npa = require('npm-package-arg')
+
+const searchCachePackage = (spec, cacheKeys) => {
+  const parsed = npa(spec)
+  if (parsed.rawSpec !== '' && parsed.type === 'tag')
+    throw new Error(`Cannot list cache keys for a tagged package.`)
+  const searchMFH = new RegExp(`^make-fetch-happen:request-cache:.*(?<!/[@a-zA-Z]+)/${parsed.name}/-/(.*)-(?<ver>.*).tgz$`)
```
Comment (wraithgar) [inline lib/cache.js]:
```javascript
> r
/\/-\/(.*)-(?<ver>.*).tgz$/
> '/-/foo-1.2.3-beta.tgz'.match(r)
[
  '/-/foo-1.2.3-beta.tgz',
  'foo-1.2.3',
  'beta',
  index: 0,
  input: '/-/foo-1.2.3-beta.tgz',
  groups: [Object: null prototype] { ver: 'beta' }
```

Comment (wraithgar) [reply lib/cache.js]: The only reliable way to pull a version out of a tarball url is to know the name beforehand and explicitly strip out `name-` from the base filename

Comment (fritzy) [reply lib/cache.js]: I made a test that failed, implemented something that extracted the version more reliably, and made sure the tests then passed.

File: lib/cache.js (hunk 2 above)
Comment (fritzy) [reply lib/cache.js]: now accepts <key>

File: lib/cache.js (hunk 1 above)
Comment (fritzy) [reply lib/cache.js]: Now uses `npm-package-arg`

File: test/lib/cache.js [hunk 5]
```diff
@@ -45,6 +45,7 @@ const setupCacacheFixture = () => {
     'webpack@4.44.1',
     'npm@1.2.0',
     'webpack@4.47.0',
+    'foo@1.2.3-beta',
```
Comment (wraithgar) [inline test/lib/cache.js]: let's also represent a package with a `-` with a semver both with and without a `-`

Comment (wraithgar) [reply test/lib/cache.js]: we should also have scoped packages in here

File: lib/cache.js [hunk 6]
```diff
@@ -4,7 +4,29 @@ const log = require('npmlog')
...
 const rimraf = promisify(require('rimraf'))
+const semver = require('semver')
 const BaseCommand = require('./base-command.js')
+const npa = require('npm-package-arg')
+
+const searchCachePackage = (spec, cacheKeys) => {
+  const parsed = npa(spec)
+  if (parsed.rawSpec !== '' && parsed.type === 'tag')
+    throw new Error(`Cannot list cache keys for a tagged package.`)
+  const searchMFH = new RegExp(`^make-fetch-happen:request-cache:.*(?<!/[@a-zA-Z]+)/${parsed.name}/-/([^/]+.tgz)$`)
```
Comment (nlf) [inline lib/cache.js]: so this regex seems right for packages in the npm registry, but i want to raise a bit of a gotcha here so we can discuss it as part of the feature..

packuments are at a known path ahead of time, that's how we discover where tarballs are, so the `searchPack` regex below here is spot on. tarballs, however, are another story. there is no mandated path pattern for where a registry keeps its tarballs, we use the tarball url stored in the packument and fetch that. so while this feature will work for public registry users (or alternative registries that follow our path scheme), it is very likely it will miss tarball urls for third party registries.

_if_ we have a packument for the package in question, we could read its contents from the cache without sending a request and gather up the urls from there and look for those. if we don't have a packument, though, we're kinda stuck without sending a request..

File: test/lib/cache.js (hunk 5 above)
Comment (fritzy) [reply test/lib/cache.js]: added tests

File: lib/cache.js (hunk 6 above)
Comment (fritzy) [reply lib/cache.js]: It now checks for keys from packument tarballs in addition to url parsing to cover registries that store packages at different paths. It still also checks for keys matching the public registry pattern.

File: lib/cache.js [hunk 7]
```diff
@@ -131,6 +201,25 @@ with --force.`)
...
+    if (Array.isArray(this.npm.flatOptions.package)
+      && this.npm.flatOptions.package.length > 0) {
+      // get results for each package spec specified
+      const results = new Set()
+      for (const spec of this.npm.flatOptions.package) {
+        const keySet = await searchCachePackage(cachePath, spec, cacheKeys)
+        for (const key of keySet)
+          results.add(key)
+      }
+      [...results].sort().forEach(key => this.npm.output(key))
```
Comment (nlf) [inline lib/cache.js]: just want to confirm: are we using a `Set` for results (and for the `cacheKeys`, which is converted to a `Set` in the `searchCachePackage` function) to avoid duplicates? do we really get duplicates?

also, when sorting package names elsewhere we generally use `localeCompare`, it feels like it would be best to follow suit here and change this to something like:

```suggestion
      [...results].sort((a, b) => a.localeCompare(b, 'en')).forEach((key) => this.npm.output(key))
```

File: lib/cache.js [hunk 8]
```diff
@@ -131,6 +201,25 @@ with --force.`)
...
+      return
+    }
+    cacheKeys.sort().forEach(key => this.npm.output(key))
```
Comment (nlf) [inline lib/cache.js]: same comment regarding the sorting here, would be best to switch this to use `localeCompare`

File: lib/cache.js [hunk 9]
```diff
@@ -4,7 +4,61 @@ const log = require('npmlog')
...
+      continue
+
+    results.add(key)
+    let packument, details
+    try {
+      details = await cacache.get(path, key)
+      packument = jsonParse(details.data)
+    } catch (_) {
+      // if we couldn't parse the packument, abort
+      /* istanbul ignore next */
```
Comment (nlf) [inline lib/cache.js]: as a general guiding principle we avoid disabling coverage unless there's an edge case that's overly difficult to reach within a test. when that happens, we also typically leave a comment explaining why reaching the line for coverage is difficult.

this case here, and the two similar ones below it, seem like they could be tested without too much difficulty

File: lib/cache.js [hunk 10]
```diff
@@ -4,7 +4,61 @@ const log = require('npmlog')
...
+    }
+    /* istanbul ignore next */
+    if (!packument.versions || typeof packument.versions !== 'object')
+      continue
+    // assuming this is a packument
+    for (const ver of Object.keys(packument.versions)) {
+      if (semver.satisfies(ver, parsed.rawSpec)) {
+        /* istanbul ignore else */
+        if (packument.versions[ver].dist
+          && typeof packument.versions[ver].dist === 'object'
```
Comment (nlf) [inline lib/cache.js:51]: so we've verified that `packument.versions[ver].dist` is an object, but what if that object doesn't have a `tarball` property? is that worth checking for here? the consequence is we'll search for `make-fetch-happen:request-cache:undefined` as a key which _could_ hit a match, but if it does that would be surprising and most likely indicative of a bug somewhere, so maybe it's ok to ignore it?

**Review by nlf — CHANGES_REQUESTED** [review]
in addition to the comments above, one last thing:

the `params` static getter should have `'package'` added to it, that'll make sure that when docs are generated it will list the `--package` flag as one in use by this command

File: lib/cache.js (hunk 7 above)
Comment (nlf) [reply lib/cache.js]: > just want to confirm: are we using a Set for results (and for the cacheKeys, which is converted to a Set in the searchCachePackage function) to avoid duplicates? do we really get duplicates?

answered elsewhere, the duplicates stem from us searching both for the specific urls listed in the packument as well as scanning all keys for the known URL pattern. we end up finding each key for a tarball in the public registry twice

File: lib/cache.js (hunk 10 above)
Comment (fritzy) [reply lib/cache.js:51]: addressed by adding a condition

File: lib/cache.js (hunk 9 above)
Comment (fritzy) [reply lib/cache.js]: I removed the code coverage exceptions and added tests to deal with corrupted or incorrect packuments.

File: lib/cache.js (hunk 8 above)
Comment (fritzy) [reply lib/cache.js]: I updated the sorts to use locale

**Review by nlf — APPROVED** [review]
looks great!
